tools/ci-operator-prowgen/main.go

```diff
@@ -0,0 +1,318 @@
+			Decorate:         true,
+		},
+	}
+	return &postsubmit
+}
+
+// Given a ci-operator configuration file and basic information about what
+// should be tested, generate a following JobConfig:
+//
+// - one presubmit and one postsubmit for each test defined in config file
+// - if the config file has non-empty `images` section, generate an additinal
+//   presubmit and postsubmit that has `--target=[images]`. This postsubmit
+//   will additionally pass `--promote` to ci-operator and it will have
+//   `artifacts: images` label
+func generateJobs(
+	configSpec *cioperatorapi.ReleaseBuildConfiguration,
+	org, repo, branch string,
+) *prowconfig.JobConfig {
+
+	orgrepo := fmt.Sprintf("%s/%s", org, repo)
+	presubmits := map[string][]prowconfig.Presubmit{}
+	postsubmits := map[string][]prowconfig.Postsubmit{}
+
+	imagesTest := false
+
+	for _, element := range configSpec.Tests {
+		// Check if config file has "images" test defined to avoid name clash
+		// (we generate the additional `--target=[images]` jobs name with `images`
+		// as an identifier, but a user can have `images` test defined in his
+		// config file which would result in a clash)
+		if element.As == "images" {
+			imagesTest = true
+		}
+		test := testDescription{Name: element.As, Target: element.As}
+		presubmits[orgrepo] = append(presubmits[orgrepo], *generatePresubmitForTest(test, org, repo, branch))
+		postsubmits[orgrepo] = append(postsubmits[orgrepo], *generatePostsubmitForTest(test, org, repo, branch))
+	}
+
+	if len(configSpec.Images) > 0 && !imagesTest {
+		// TODO: somehow handle the images case better than just not creating this job when there is name conflict
+		test := testDescription{Name: "images", Target: "[images]"}
+		presubmits[orgrepo] = append(presubmits[orgrepo], *generatePresubmitForTest(test, org, repo, branch))
+		imagesPostsubmit := generatePostsubmitForTest(test, org, repo, branch, "--promote")
+		imagesPostsubmit.Labels = map[string]string{"artifacts": "images"}
+		postsubmits[orgrepo] = append(postsubmits[orgrepo], *imagesPostsubmit)
+	}
+
+	return &prowconfig.JobConfig{
+		Presubmits:  presubmits,
+		Postsubmits: postsubmits,
+	}
+}
+
+func readCiOperatorConfig(configFilePath string) *cioperatorapi.ReleaseBuildConfiguration {
+	data, err := ioutil.ReadFile(configFilePath)
+	if err != nil {
+		fmt.Fprintf(os.Stderr, "failed to read ci-operator config (%v)\n", err)
+		os.Exit(1)
+	}
+
+	var configSpec *cioperatorapi.ReleaseBuildConfiguration
+	if err := json.Unmarshal(data, &configSpec); err != nil {
+		fmt.Printf("failed to load ci-operator config (%v)\n", err)
+		os.Exit(1)
+	}
+
+	return configSpec
+}
+
+// We use the directory/file naming convention to encode useful information
+// about component repository information.
+// The convention for ci-operator config files in this repo:
+// ci-operator/config/ORGANIZATION/COMPONENT/BRANCH.json
+func extractRepoElementsFromPath(configFilePath string) (string, string, string) {
```

Comment:
What happens in error cases?

Reply:
This should happen only when someone passes some funky path to `--source-config`, which we warn against in [README](https://github.com/openshift/release/pull/1160/commits/25065fd8a9bea36569b15f1517bf07c3cc93268a#diff-35a06896ff3190034afcf280abb5efbaR57) - we only write to stdout so at least we do not write some crap to job directories. Or someone passes a path to `--config-dir ../../ci-operator/config/` where that path has some different directory structure but still contains ci-operator config files. I think that someone sort of knows what they are doing and can verify what was generated.

Good case, those fragments will not be there, we can detect (we do not now) it and error out.
Bad case, those fragments will be there but will be something bogus, and we will generate some crap. We could only detect this by asking github whether that org/repo:branch triplet is sane but IMO that is overkill.